src/main/java/com/yelp/nrtsearch/server/luceneserver/search/sort/SortParser.java

```diff
@@ -125,6 +129,78 @@ public static Map<String, CompositeFieldValue> getAllSortedValues(
     return values;
   }
 
+  public static FieldDoc parseLastHitInfo(
+      LastHitInfo lastHitInfo, QuerySortField querySortField, Map<String, FieldDef> queryFields) {
+    Object[] values = new Object[lastHitInfo.getLastFieldValuesCount()];
+    for (int i = 0; i < values.length; i++) {
+      SortType sortType = querySortField.getFields().getSortedFields(i);
+      try {
+        values[i] =
+            convertStringSortValue(sortType, lastHitInfo.getLastFieldValues(i), queryFields);
+      } catch (SearchHandler.SearchHandlerException e) {
+        throw new IllegalArgumentException(e);
+      }
+    }
+    return new FieldDoc(lastHitInfo.getLastDocId(), lastHitInfo.getLastScore(), values);
+  }
+
+  private static Object convertStringSortValue(
+      SortType field, String stringValue, Map<String, FieldDef> queryFields)
+      throws SearchHandler.SearchHandlerException {
+    Object sortValue;
+    FieldDef fd = null;
+    SortField.Type type;
+    if (field.getFieldName().equals(DOCID)) {
+      type = SortField.Type.DOC;
+    } else if (field.getFieldName().equals(SCORE)) {
+      type = SortField.Type.SCORE;
+    } else {
+      fd = queryFields.get(field.getFieldName());
+      if (fd == null) {
+        throw new SearchHandler.SearchHandlerException(
+            String.format(
+                "field: %s was not registered and was not specified as a virtualField",
+                field.getFieldName()));
+      }
+
+      if (!(fd instanceof Sortable)) {
+        throw new IllegalArgumentException(
+            String.format("field: %s does not support sorting", field.getFieldName()));
+      }
+      type = ((Sortable) fd).getSortField(field).getType();
+    }
+    switch (type) {
+      case DOC:
+      case INT:
+        sortValue = Integer.valueOf(stringValue);
+        break;
+      case SCORE:
+      case FLOAT:
+        sortValue = Float.valueOf(stringValue);
+        break;
+      case LONG:
+        sortValue = Long.valueOf(stringValue);
+        break;
+      case DOUBLE:
+        sortValue = Double.valueOf(stringValue);
+        break;
+      case STRING:
+      case STRING_VAL:
+        sortValue = stringValue;
+        break;
+      case CUSTOM:
+        if (fd instanceof VirtualFieldDef || fd instanceof LatLonFieldDef) {
+          sortValue = Double.valueOf(stringValue);
+        } else {
+          sortValue = stringValue;
+        }
```

Comment:
It might be good to add something to the `Sortable` interface to let the fields parse the custom values